core/src/nucleus/actions/call_zome_function.rs

```diff
@@ -141,32 +117,162 @@ pub async fn call_zome_function(
     })
 }
 
-// TODO: check the signature too
-fn is_token_the_agent(context: Arc<Context>, cap: &Option<CapabilityCall>) -> bool {
-    match cap {
-        None => false,
-        Some(call) => context.agent_id.key == call.cap_token.to_string(),
+/// validates that a given zome function call specifies a correct zome function and capability grant
+pub fn validate_call(
+    context: Arc<Context>,
+    fn_call: &ZomeFnCall,
+) -> Result<(String, DnaWasm), HolochainError> {
+    let state = context.state().ok_or(HolochainError::ErrorGeneric(
+        "Context not initialized".to_string(),
+    ))?;
+
+    let nucleus_state = state.nucleus();
+
+    // make sure the dna, zome and function exists and return pretty errors if they don't
+    let dna = nucleus_state
+        .dna()
+        .ok_or_else(|| HolochainError::DnaMissing)?;
+    let zome = dna
+        .get_zome(&fn_call.zome_name)
+        .map_err(|e| HolochainError::Dna(e))?;
+    let _ = dna
+        .get_function_with_zome_name(&fn_call.zome_name, &fn_call.fn_name)
+        .map_err(|e| HolochainError::Dna(e))?;
+
+    if check_capability(context.clone(), fn_call)
+        || (is_token_the_agent(context.clone(), &fn_call.cap)
+            && verify_call_sig(
+                context.clone(),
+                &fn_call.cap.provenance.1,
+                &fn_call.fn_name,
+                fn_call.parameters.clone(),
+            ))
+    {
+        Ok((dna.name.clone(), zome.code.clone()))
+    } else {
+        Err(HolochainError::CapabilityCheckFailed)
     }
 }
 
-/// checks to see if a given function call is allowable according to the capabilities
-/// that have been registered to callers in the chain.
-fn check_capability(context: Arc<Context>, fn_call: &ZomeFnCall) -> bool {
-    // the agent can always do everything
-    if is_token_the_agent(context.clone(), &fn_call.cap) {
-        return true;
+fn is_token_the_agent(context: Arc<Context>, request: &CapabilityRequest) -> bool {
+    context.agent_id.key == request.cap_token.to_string()
```

Comment:
Should this be:
```suggestion
    context.agent_id.address()? == request.cap_token.to_string()
```
?